### PR TITLE
Btrx 666 dgil

### PR DESCRIPTION
**Pull request by dags**
## Addresses BTRX-666 - Configure ORSP GKE deployment... (DO NOT MERGE YET)
https://broadinstitute.atlassian.net/browse/BTRX-666

## Changes
- Circleci configuration files were changed.
- location and reading of orsp-client.json file was modified.

## Testing
Application should be deployed to https://ui.dsp-orsp-dev.broadinstitute.org/ and work as usual. 

---

- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Get a thumb from Belatrix
- [ ] **Submitter**: Get a thumb from @rushtong
- [ ] **Submitter**: Merge to develop using github's "Squash and Merge" feature
- [ ] **Submitter**: Delete branch after merge

### CODE REVIEW

File: conf/ingress.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  annotations:
+    kubernetes.io/ingress.global-static-ip-name: orsp-pub-dev-ip
+  labels:
+    app: ${REPO_NAME}
+  name: ${REPO_NAME}-ingress
+  namespace: default
```
Comment (rushtong) [inline conf/ingress.yaml]: We don't recommend the default namespace so that the cluster could be used for other namespaced apps.

File: conf/ingress.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  rules:
+  - host: ui.dsp-orsp-dev.broadinstitute.org
+    http:
+      paths:
+      - backend:
+          serviceName: ${REPO_NAME}-service
+          servicePort: http
+  tls:
+  - hosts:
+    - ui.dsp-orsp-dev.broadinstitute.org
```
Comment (rushtong) [inline conf/ingress.yaml]: How permanent is this domain name? Who manages `dsp-orsp-dev` ?

File: conf/k8s.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      restartPolicy: Always
+      schedulerName: default-scheduler
+      securityContext: {}
+      terminationGracePeriodSeconds: 30
+      volumes:
+      - name: cloudsql-instance-credentials
+        secret:
+          secretName: cloudsql-instance-credentials
+      - configMap:
+          name: orsp-config-dev
```
Comment (rushtong) [inline conf/k8s.yml]: Should this be parameterized? It won't work for any other environment.

File: conf/k8s.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        imagePullPolicy: Always
+        name: ${REPO_NAME}
+        ports:
+        - containerPort: 8080
+        volumeMounts:
+        - mountPath: "/orsp-config"
+          name: orsp-config
+      - command:
+        - "/cloud_sql_proxy"
+        - "-instances=orsp-dev:us-east1:bricc-mysql-dev=tcp:3306"
```
Comment (rushtong) [inline conf/k8s.yml]: I think this should be injected based n environment.

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,16 @@
+FROM redwolfgang20/grails as builder
+COPY . /app
+WORKDIR /app
+
+RUN ./gradlew webpackProd
+RUN grails -Dgrails.env=dev war
```
Comment (rushtong) [inline Dockerfile]: This doesn't seem right ... env=dev?

Comment (rushtong) [reply Dockerfile]: Following up on this. We do need to know how the image should be run, and that might be done as an environment variable in the cluster rather than hard coding it here. We do need to distinguish dev/prod/etc. so we don't spam real users with email, etc.

**Comment by rushtong** [comment]
Is the new deploy code using anything from `buildSrc`? If not, I suggest reverting all of that code and leave it as a tech-debt task to remove all of it.

File: conf/ingress.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  labels:
+    app: ${REPO_NAME}
+  name: ${REPO_NAME}-ingress
+  namespace: default
+spec:
+  backend:
+    serviceName: ${REPO_NAME}-service
+    servicePort: http
+  rules:
+  - host: ui.dsp-orsp-dev.broadinstitute.org
```
Comment (rushtong) [inline conf/ingress.yaml]: Shouldn't this be parameterized somehow?

File: conf/ingress.yaml (hunk 1 above)
Comment (dags) [reply conf/ingress.yaml]: Got it. I'll move all to an orsp namespace.

File: conf/ingress.yaml (hunk 2 above)
Comment (dags) [reply conf/ingress.yaml]: It is as permanent as we wish, I guess, dsp-orsp-dev DSN Zone (Cloud DNS) is under control of orsp-pub project, us, in a similar way as dsp-duos-dev is under broad-duos-dev project. That is part of the work done by Henry Ferrrara's team after my trip to Boston.

File: conf/k8s.yml (hunk 3 above)
Comment (dags) [reply conf/k8s.yml]: Yes, we should create a config map for every environment. I'll fix it.

File: conf/k8s.yml (hunk 4 above)
Comment (dags) [reply conf/k8s.yml]: Options: instance name may be the same if we have a database per environment in same mysql instance. Or we may have N mysql instances, one per environment,  with just one database (same name) per instance. Not sure what may be the best option to choose. Thoughts ?

File: Dockerfile (hunk 5 above)
Comment (dags) [reply Dockerfile]: Yes, you are right. I used this just as a workaround to exiting build process: requires a dev or prod parameter. Actually, it doesn't matter much, we could use prod as well. Build process should be environment agnostic, I guess, to create a single docker image shared among all environments. Actual environment will be configured through the ConfigMap mounted in runtime. I'll take a look at build process to remove the need of a env parameter.

File: conf/ingress.yaml (hunk 6 above)
Comment (dags) [reply conf/ingress.yaml]: yes. Will find where and add it.

File: conf/k8s.yml (hunk 4 above)
Comment (rushtong) [reply conf/k8s.yml]: > Options: instance name may be the same if we have a database per environment in same mysql instance. Or we may have N mysql instances, one per environment, with just one database (same name) per instance. Not sure what may be the best option to choose. Thoughts ?

My suggestion is to have completely separate database instances, each uniquely named, one per cluster that is configured just for that cluster. If we want to have more envs in the project, we can, and each one can point to their own db instance. This way, you can have a `dev` cluster, a `dgil-dev` cluster for your own dev work, a `grushton-dev` cluster for my own work, etc. It also frees us up to have a `staging` or `perf` environment in the same google project.

In that world, the database name is a configuration and not hard-coded.

File: build.gradle [hunk 7]
```diff
@@ -217,31 +224,31 @@ task ('cleanTest') {
 task renderConfigs(type: RenderConfigs) {
     local = false
     projectHome = projectDir
-    vault = System.getenv('VAULT_ADDR')
+    vaultAddr = System.getenv('VAULT_ADDR')
```
Comment (rushtong) [inline build.gradle]: This change seems unnecessary to me. In the current model, all configurations are in Circleci and rendered out via circleci.yml. If that is the route we are taking, then this task should be deleted, along with *all* of the code from `buildSrc`, which includes the `RenderConfigs` class.

File: build.gradle [hunk 8]
```diff
@@ -217,31 +224,31 @@ task ('cleanTest') {
...
+//     appPath = env.equals(Deploy.DeployEnv.prod.name()) ? '/' : "/${env}"
+//     host = 'https://orsp.broadinstitute.org' // Overrideable option
+//     user = ''   // Required option
+//     pass = ''   // Required option
+//     dependsOn renderConfigs
+//     dependsOn webpackProd
+//     dependsOn war
+//     tasks.findByName('war').mustRunAfter webpackProd
+//     tasks.findByName('war').mustRunAfter renderConfigs
+// }
```
Comment (rushtong) [inline build.gradle]: Delete.

File: buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/RenderConfigs.groovy [hunk 9]
```diff
@@ -12,25 +12,22 @@ class RenderConfigs extends DefaultTask {
     void setLocalConfig(boolean localConfig) { this.local = localConfig }
 
     public String projectHome
-    String vault
+    String vaultAddr
```
Comment (rushtong) [inline buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/RenderConfigs.groovy]: delete this class file

File: buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/RenderTemplate.groovy [hunk 10]
```diff
@@ -26,7 +26,8 @@ class RenderTemplate {
                 dropRight(1).       // Drop the last line which ends the consul variables
                 join("\n")          // Turn it back into a string.
         // TODO: This should be an input
-        new File("${projectHome}/grails-app/conf/application.yml").withWriter { w ->
+        new File("/etc/config/application.yml").withWriter { w ->
+        //new File("${projectHome}/grails-app/conf/application.yml").withWriter { w ->
```
Comment (rushtong) [inline buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/RenderTemplate.groovy]: delete this class file

File: buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/VaultSecrets.groovy [hunk 11]
```diff
@@ -7,12 +7,14 @@ import groovyx.net.http.HttpBuilder
 class VaultSecrets {
 
     static Map<String, String> getParsedSecret(String token, String vaultAddress, String secretPath) {
+
```
Comment (rushtong) [inline buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/VaultSecrets.groovy]: delete this class file

File: grails-app/controllers/org/broadinstitute/orsp/SearchController.groovy [hunk 12]
```diff
@@ -195,7 +195,7 @@ class SearchController implements UserInfo {
         if (params.dateRestriction) options.dateRestriction = Date.parse('mm/dd/yyyy', params.dateRestriction)
         if (params.methodsResearchExcluded) options.methodsResearchExcluded = true
         List<Issue> issues = queryService.findByQueryOptions(options)
-        render (view: '_dataUseRecords', model: ["issues": issues])
+        render(view: '_dataUseRecords', model: ["issues": issues])
```
Comment (rushtong) [inline grails-app/controllers/org/broadinstitute/orsp/SearchController.groovy]: All changes to this class should be reverted

File: grails-app/services/org/broadinstitute/orsp/CrowdService.groovy [hunk 13]
```diff
@@ -1,131 +1,118 @@
...
+import com.google.cloud.bigquery.BigQuery
+import com.google.cloud.bigquery.BigQueryOptions
+import com.google.cloud.bigquery.FieldValueList
+import com.google.cloud.bigquery.Job
+import com.google.cloud.bigquery.JobId
+import com.google.cloud.bigquery.JobInfo
+import com.google.cloud.bigquery.QueryJobConfiguration
+import com.google.cloud.bigquery.TableResult
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.auth.oauth2.ServiceAccountCredentials;
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/CrowdService.groovy]: All changes to this class should be reverted and put into a separate PR that just addresses BTRX-600. There should be no code that calls `CrowdService` since that is handled via a cronjob on the current Broad VM.

File: grails-app/services/org/broadinstitute/orsp/PermissionService.groovy [hunk 14]
```diff
@@ -46,4 +46,3 @@ class PermissionService implements UserInfo {
         !userHasAccess
     }
 }
-
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: Revert

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy [hunk 15]
```diff
@@ -1321,20 +1351,20 @@ class QueryService implements Status {
...
+            sqlQuery.setLong("version", 0)
+            sqlQuery.setString("role", it)
+            sqlQuery.setString("userName", user.userName)
+            sqlQuery.setLong("userId", user.id)
+            sqlQuery.executeUpdate()
+        }
     }
 
-    void deleteOrspUserRoles (userId) {
+    void deleteOrspUserRoles(userId) {
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: Revert all changes to this class - they are not deployment related. Feel free to put up a separate PR that addresses what you want these changes to accomplish.

File: grails-app/services/org/broadinstitute/orsp/StorageProviderService.groovy [hunk 16]
```diff
@@ -441,10 +441,9 @@ class StorageProviderService implements Status {
      * @return A GoogleCredential from json secrets.
      */
     private GoogleCredential getCredential() {
+
         if (!credential) {
-            File configFile = GrailsResourceUtils.getFile(
-                    GrailsResourceUtils.toURI(this.getClass().getResource(storageConfiguration.config)),
-                    "GCS Config")
+            File configFile = new File(storageConfiguration.config);
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/StorageProviderService.groovy]: Question: Is this necessary for deployment? If not, revert and file a separate PR for the simplification.

File: grails-app/views/layouts/_footer.gsp [hunk 17]
```diff
@@ -37,7 +37,7 @@
 <asset:javascript src="jquery.fn.dataTablesExt.ticket.js"/>
 <asset:javascript src="chosen.jquery.min.js"/>
 <asset:javascript src="jasny-bootstrap.min.js"/>
-<script src="https://cloud.tinymce.com/5/tinymce.min.js?apiKey=8zknubfnjvv9l3sg0cpxrome1qk6r2wlpdw7j4ebb3gjxige"></script>
+<script src="https://cloud.tinymce.com/5/tinymce.min.js?apiKey=kyknyn3xmeam53u8vyway2oxf18oecyfwkjoym1xcydpyfyh"></script>
```
Comment (rushtong) [inline grails-app/views/layouts/_footer.gsp]: Is this also necessary for the version deployed to the Broad VM or is it only for the cloud version? Also, is it necessary at all? I thought that Leo submitted a PR to replace this.

File: grails-app/views/layouts/main_component.gsp [hunk 18]
```diff
@@ -147,7 +147,7 @@
     <script src="//cdn.datatables.net/buttons/1.5.1/js/buttons.print.min.js"></script>
     <script src="//cdnjs.cloudflare.com/ajax/libs/moment.js/2.8.4/moment.min.js"></script>
     <script src="//cdn.datatables.net/plug-ins/1.10.19/sorting/datetime-moment.js"></script>
-    <script src="https://cloud.tinymce.com/5/tinymce.min.js?apiKey=8zknubfnjvv9l3sg0cpxrome1qk6r2wlpdw7j4ebb3gjxige"></script>
+    <script src="https://cloud.tinymce.com/5/tinymce.min.js?apiKey=kyknyn3xmeam53u8vyway2oxf18oecyfwkjoym1xcydpyfyh"></script>
```
Comment (rushtong) [inline grails-app/views/layouts/main_component.gsp]: Same question here as above.

File: package.json [hunk 19]
```diff
@@ -45,7 +45,9 @@
     "babel-polyfill": "6.26.0",
     "babel-register": "6.26.0",
     "bootstrap": "4.3.1",
+    "classnames": "^2.2.6",
     "date-fns": "1.30.1",
+    "jquery": "^1.9.1",
```
Comment (rushtong) [inline package.json]: How are these changes necessary for deployment? Revert if not.

File: src/integration-test/groovy/org/broadinstitute/orsp/CrowdServiceSpec.groovy [hunk 20]
```diff
@@ -32,49 +32,49 @@ class CrowdServiceSpec extends Specification {
     def cleanup() {
     }
 
-    void stubCrowdResource() {
```
Comment (rushtong) [inline src/integration-test/groovy/org/broadinstitute/orsp/CrowdServiceSpec.groovy]: I suggest disabling these tests for this PR and removing it entirely for BTRX-600 (and replacing it with tests that cover the BQ implementation)

File: src/main/groovy/org/broadinstitute/orsp/IssueSearchItemDTO.groovy [hunk 21]
```diff
@@ -19,7 +19,6 @@ class IssueSearchItemDTO {
     Map<String, Set<String>> contactNames = new HashMap<String, Set<String>>()
     Map<String, Set<String>> extraProperties = new HashMap<String, Set<String>>()
 
-
```
Comment (rushtong) [inline src/main/groovy/org/broadinstitute/orsp/IssueSearchItemDTO.groovy]: revert

File: src/main/groovy/org/broadinstitute/orsp/config/CrowdConfiguration.groovy [hunk 22]
```diff
@@ -6,9 +6,5 @@ import org.springframework.context.annotation.Configuration
 @Configuration
 @ConfigurationProperties(prefix = "crowd")
 class CrowdConfiguration {
-    String url
-    String username
-    String password
-    Integer interval
-    Integer range
+    String config
```
Comment (rushtong) [inline src/main/groovy/org/broadinstitute/orsp/config/CrowdConfiguration.groovy]: Revert and address in BTRX-600

File: src/main/webapp/search/Search.js [hunk 23]
```diff
@@ -150,6 +150,7 @@ class Search extends React.Component {
     this.state.irb.map(function(irb, index) {
       params.append("irb", irb.id);
     });
+
```
Comment (rushtong) [inline src/main/webapp/search/Search.js]: Revert

File: src/test/groovy/org/broadinstitute/orsp/config/CrowdConfigurationSpec.groovy [hunk 24]
```diff
@@ -15,11 +15,7 @@ class CrowdConfigurationSpec extends Specification implements ServiceUnitTest<Cr
         CrowdConfiguration config = service.crowdConfiguration
 
         expect:
-        !config.url.isEmpty()
-        !config.username.isEmpty()
-        !config.password.isEmpty()
-        config.interval > 0
-        config.range > 0
+        !config.url.config()
```
Comment (rushtong) [inline src/test/groovy/org/broadinstitute/orsp/config/CrowdConfigurationSpec.groovy]: Revert and address in BTRX-600

**Review by rushtong — CHANGES_REQUESTED** [review]
I know this is still WIP, but it needs to be expedited ASAP so all new code is auto-deployed and immediately visible for ORSP testing.

File: grails-app/views/layouts/_footer.gsp (hunk 17 above)
Comment (rushtong) [reply grails-app/views/layouts/_footer.gsp]: Answering my own question, we're using an updated version of tinymce (https://github.com/tinymce/tinymce-react) so we do still need an API key with this PR: https://github.com/broadinstitute/orsp-pub/pull/208

File: build.gradle (hunk 7 above)
Comment (dags) [reply build.gradle]: done

File: buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/RenderConfigs.groovy (hunk 9 above)
Comment (dags) [reply buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/RenderConfigs.groovy]: done

File: buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/RenderTemplate.groovy (hunk 10 above)
Comment (dags) [reply buildSrc/src/main/groovy/org/broadinstitute/orsp/deploy/RenderTemplate.groovy]: done

File: grails-app/controllers/org/broadinstitute/orsp/SearchController.groovy (hunk 12 above)
Comment (dags) [reply grails-app/controllers/org/broadinstitute/orsp/SearchController.groovy]: done

File: grails-app/services/org/broadinstitute/orsp/CrowdService.groovy (hunk 13 above)
Comment (dags) [reply grails-app/services/org/broadinstitute/orsp/CrowdService.groovy]: done

File: grails-app/services/org/broadinstitute/orsp/QueryService.groovy (hunk 15 above)
Comment (dags) [reply grails-app/services/org/broadinstitute/orsp/QueryService.groovy]: done

File: grails-app/services/org/broadinstitute/orsp/StorageProviderService.groovy (hunk 16 above)
Comment (dags) [reply grails-app/services/org/broadinstitute/orsp/StorageProviderService.groovy]: I think it is, to get configuration for storage.

File: package.json (hunk 19 above)
Comment (dags) [reply package.json]: This fix a complaint about missing jquery, from bootstrap. it seems to work without it anyway but keeps complaining.

File: src/main/groovy/org/broadinstitute/orsp/IssueSearchItemDTO.groovy (hunk 21 above)
Comment (dags) [reply src/main/groovy/org/broadinstitute/orsp/IssueSearchItemDTO.groovy]: done

File: src/main/webapp/search/Search.js (hunk 23 above)
Comment (dags) [reply src/main/webapp/search/Search.js]: done

File: src/test/groovy/org/broadinstitute/orsp/config/CrowdConfigurationSpec.groovy (hunk 24 above)
Comment (dags) [reply src/test/groovy/org/broadinstitute/orsp/config/CrowdConfigurationSpec.groovy]: done

File: src/main/groovy/org/broadinstitute/orsp/config/CrowdConfiguration.groovy (hunk 22 above)
Comment (dags) [reply src/main/groovy/org/broadinstitute/orsp/config/CrowdConfiguration.groovy]: done

File: grails-app/services/org/broadinstitute/orsp/PermissionService.groovy (hunk 14 above)
Comment (dags) [reply grails-app/services/org/broadinstitute/orsp/PermissionService.groovy]: done

File: Dockerfile (hunk 5 above)
Comment (rushtong) [reply Dockerfile]: This is still outstanding issue.

File: Dockerfile [hunk 25]
```diff
@@ -0,0 +1,16 @@
+FROM redwolfgang20/grails as builder
```
Comment (rushtong) [inline Dockerfile]: This makes me nervous. It's entry in dockerhub does not look like a production quality base image.

File: conf/k8s.yml (hunk 4 above)
Comment (rushtong) [reply conf/k8s.yml]: This is still an outstanding problem. The db name should be parameterized somehow.

File: package.json (hunk 19 above)
Comment (rushtong) [reply package.json]: Please file it as a separate PR. Anyone on the team can take on these kinds of minor updates that are unrelated to the task of auto-deployment.

File: src/main/config/application.yml.ctmpl [hunk 26]
```diff
@@ -67,10 +67,10 @@ environments:
       password: {{$orsp.Data.crowd_password}}
       interval: 500
       range: 10
-    grails.serverURL: https://orsp.broadinstitute.org/dev
+    grails.serverURL: https://ui.dsp-orsp-dev.broadinstitute.org
     googleSignInClientId: {{$orsp.Data.google_signin_client_id}}
```
Comment (rushtong) [inline src/main/config/application.yml.ctmpl]: How are these getting substituted? I don't see that anywhere in this PR.

File: webpack.config.js [hunk 27]
```diff
@@ -14,6 +14,7 @@ module.exports = {
     linkWizard: './src/main/webapp/linkWizard/index.js',
     mainIndex: './src/main/webapp/main/main_index.js',
     search: './src/main/webapp/search/index.js',
+    project: './src/main/webapp/project/index.js'
```
Comment (rushtong) [inline webpack.config.js]: Why isn't this in feature already? It's not related to deployment ... it should already be part of the main build, but it doesn't look to be the case now. Should be a separate PR.

**Review by rushtong — CHANGES_REQUESTED** [review]
See comments and questions inline.

File: src/main/config/application.yml.ctmpl (hunk 26 above)
Comment (dags) [reply src/main/config/application.yml.ctmpl]: I do not understand this one.

File: webpack.config.js (hunk 27 above)
Comment (dags) [reply webpack.config.js]: You are right. I thought it was. Moving it to a separated PR.

File: src/main/config/application.yml.ctmpl (hunk 26 above)
Comment (rushtong) [reply src/main/config/application.yml.ctmpl]: The value of `{{$orsp.Data.google_signin_client_id}}` is currently pulled from vault secrets and rendered during the conversion of the ctmpl file to the yml file. I'm guessing that there is a very different mechanism for doing the rendering, but I haven't looked at the latest documentation.

First question: Is this file being used any more? If not, delete it. 
Second question: If this file is being used, where is `$orsp.Data.google_signin_client_id` coming from?

Comment (dags) [reply src/main/config/application.yml.ctmpl]: got it. Will delete it.

File: Dockerfile (hunk 25 above)
Comment (dags) [reply Dockerfile]: replaced with an alternative approach.

File: conf/k8s.yml (hunk 4 above)
Comment (dags) [reply conf/k8s.yml]: I think this the last config change pending. I am on, working on proper syntax.

File: .circleci/config.yml [hunk 28]
```diff
@@ -1,50 +1,213 @@
...
+        type: string
+        default: "orsp-dev"
+    steps:
+      - run:
+          name: Export env vars for substitution. $BASH_ENV is sourced before each step
+          command: |
+            echo 'export COMMIT="`git rev-parse --short HEAD`"' >> $BASH_ENV
+            echo 'export ENV="<< parameters.env >>"' >> $BASH_ENV
+            echo 'export REPO_NAME="<< parameters.repo_name >>"' >> $BASH_ENV
+            echo 'export REPO_BRANCH="<< parameters.repo_branch >>"' >> $BASH_ENV
```
Comment (rushtong) [inline .circleci/config.yml]: Do we actually use this anywhere?

File: Dockerfile [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+
+#
+
+RUN curl -s "https://get.sdkman.io" | bash
+RUN chmod a+x "$HOME/.sdkman/bin/sdkman-init.sh"
+RUN /bin/bash -c "source $HOME/.sdkman/bin/sdkman-init.sh"
+RUN export SDKMAN_DIR="$HOME/.sdkman"
+RUN /bin/bash -c "source $HOME/.sdkman/bin/sdkman-init.sh; sdk install grails 3.3.8"
+RUN /bin/bash -c "source $HOME/.sdkman/bin/sdkman-init.sh; sdk install gradle 3.5"
+RUN /bin/bash -c "source $HOME/.sdkman/bin/sdkman-init.sh; sdk install groovy 2.5.7"
```
Comment (rushtong) [inline Dockerfile]: nit: I think all of these run commands could be one-lined and reduce the image size by a huge amount, but that would be an optimization that could go into a follow-up PR.

**Review by rushtong — APPROVED** [review]
One tiny question and a suggestion for a follow-up PR, but this looks good! 👍

**Review by vvicario — APPROVED** [review]

File: .circleci/config.yml (hunk 28 above)
Comment (dags) [reply .circleci/config.yml]: not yet, was intended to be used in env selection. Will remove if not used in a next PR.

File: Dockerfile (hunk 29 above)
Comment (dags) [reply Dockerfile]: Makes sense. This is on the builder image, should not impact on final image, I think. Will do it in a next PR.
